cpp/grammar_functor.h

```diff
@@ -337,14 +347,15 @@ class GrammarFSMBuilder {
   static std::optional<FSMWithStartEnd> Choices(const GrammarExpr& expr, const Grammar& grammar);
   static std::optional<FSMWithStartEnd> TagDispatch(const Grammar::Impl::TagDispatch& tag_dispatch);
 };
-class SubGrammarAdder {
+
+class RepetitionNormalizer {
  public:
-  static int32_t Apply(GrammarBuilder* builder, const Grammar& sub_grammar);
+  static void Apply(Grammar& grammar);
 };
 
-class RepetitionNormalizer {
+class GrammarOptimizer {
```

Comment:
add docs. expecially you should mention: if the grammar is already optimized, we will still optimize it; a new constructed grammar is returned.